### PR TITLE
logx zknova points adapter

### PR DESCRIPTION
**Pull request by arpitsingh1409**

### CODE REVIEW

**Comment by xsteadybcgo** [comment]
@arpitsingh1409 
image
For this [transaction](https://explorer.zklink.io/tx/0xc0c7520eff73efa5ee4c45b219312a0b0cd67100a5decd8064a287c2ae4f69e9), can you describe how I should verify this quantity? Is it a swap transaction? Which event can I use to verify it?

**Comment by xsteadybcgo** [comment]
CollateralDelta refers to the amount of collateral, and sizeDelta refers to the total amount with leverage applied, correct?

**Comment by arpitsingh1409** [comment]
> CollateralDelta refers to the amount of collateral, and sizeDelta refers to the total amount with leverage applied, right? @arpitsingh1409

Yes, exactly.
@xsteadybcgo

**Comment by xsteadybcgo** [comment]
Great, The transaction data makes sense. @arpitsingh1409 

---

I want to check this user's balance for tvl data. The address 0x46623d659cFDEf15cf5754C56bEc3F3f0B508fd8 is the LLP address. Actually, our points are calculated based on the underlying token TVL (for logx, the underlying token might be USDT). Can you tell me the process of minting LLP? Is it obtained by supplying USDT to the contract at 0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870? What is the relationship between LLP and fLlp? Are they minted at a 1:1 ratio?

The principle of our TVL points is as follows. Let me give an example: Many users supply/deposit their assets (let's assume USDT) into a contract. All assets are locked in the LockContract (it might be the same contract where users deposit/supply their assets, it doesn't matter). Meanwhile, the contract should mint lpTokens, which could be LLP or fLlp for LogX. When calculating data by block height, you need to calculate the total LLP or fLlp minted by the contract at that block height. Suppose you want to assign points to users holding LLP. If 20,000 LLP are minted in total, and User A holds 100 LLP, which he mints by supplying USDT, and the LockContract has a total of 500,000 USDT locked, then the balance = 100 / 20,000 * 500,000 = 2,500.

I hope I have described this clearly.

I want to confirm whether LogX locks USDT into the contract at 0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870, representing the TVL (500,000 in the example)? What is the relationship between LLP and fLlp? Are they exchanged at a 1:1 ratio? Is the total supply of LLP (20,000 in the example) supplied by 0x328A7Cb7828198DCd4Fa3b62D2E2dC900b7Cf3d5? (20,000 in the example)? Our points are allocated based on the TVL of the contract's liquidity assets, and we are performing a redistribution process.

---

image

**Comment by arpitsingh1409** [comment]
> 0x328A7Cb7828198DCd4Fa3b62D2E2dC900b7Cf3d5

@xsteadybcgo 
The LLP contract address is "0x46623d659cFDEf15cf5754C56bEc3F3f0B508fd8" and a user deposits USDT to "0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870" and LLP is minted for the user. 

fLLP is basically minted when a user stakes this minted LLP in LogX and it's minted in 1:1 ratio. 
We can just take into consideration LLP here and not fLLP as a user can mint LLP and decide not to stake it, hence having LLP but fLLP.

For LLP, we can get total minted till a block height as "0x46623d659cFDEf15cf5754C56bEc3F3f0B508fd8" is an ERC20 and the total USDT stored would be USDT balance of "0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870".

Hope it makes things clear.

However, if you want I can change our backend API to return the USDT amount a user minted upto a block height instead of LLP. Please let me know.

**Comment by xsteadybcgo** [comment]
> For LLP, we can get total minted till a block height as "0x46623d659cFDEf15cf5754C56bEc3F3f0B508fd8" is an ERC20 and the total USDT stored would be USDT balance of "0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870".

@arpitsingh1409  Got it. Let me explain the points distribution rules. The total points are allocated based on the amount of USDT locked, and individual points are then distributed to users according to the proportion of LLP they hold. So, we need you to adjust the balance calculation formula. 

Taking the data from the screenshot as an example:

First, get the total amount of LLP minted by all users up at the certain block height (TotalLLP), and then get the total amount of USDT owned by `0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870` at this block height (TotalUSDT). If a user holds 49.4604971411... LLP, then the balance for calculating points should be `49.4604971411 / TotalLLP * TotalUSDT`.

Additionally, for the balance we need, please provide the raw data. It should be multiplied by 1e6.

**Comment by arpitsingh1409** [comment]
> > For LLP, we can get total minted till a block height as "0x46623d659cFDEf15cf5754C56bEc3F3f0B508fd8" is an ERC20 and the total USDT stored would be USDT balance of "0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870".
> 
> @arpitsingh1409 Got it. Let me explain the points distribution rules. The total points are allocated based on the amount of USDT locked, and individual points are then distributed to users according to the proportion of LLP they hold. So, we need to adjust the balance calculation formula.
> 
> Taking the data from the screenshot as an example:
> 
> First, get the total amount of LLP minted by all users up at the certain block height (TotalLLP), and then get the total amount of USDT owned by `0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870` at this block height (TotalUSDT). If a user holds 49.4604971411... LLP, then the balance for calculating points should be `49.4604971411 / TotalLLP * TotalUSDT`.
> 
> Additionally, for the balance we need, please provide the raw data. It should be multiplied by 1e6.

I have updated our backend api and made the necessary changes.

File: src/adapters/logx/execution/src/index.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    return {
+      timestamp: Math.floor(new Date(item.timestamp).getTime() / 1000),
+      userAddress: item.account,
+      contractAddress: contractAddress,
+      tokenAddress: tokenAddress,
+      decimals: decimals,
+      price: price,
+      quantity: Number(item.sizeDelta)/1e30,
+      txHash: item.txhash,
+      nonce: item.blockNumber,
```
Comment (xsteadybcgo) [inline src/adapters/logx/execution/src/index.js:42]: Please refer to the documentation. The nonce should be a unique value.

Comment (xsteadybcgo) [reply src/adapters/logx/execution/src/index.js:42]: No other issues 🎉

Comment (arpitsingh1409) [reply src/adapters/logx/execution/src/index.js:42]: > Please refer to the documentation. The nonce should be a unique value.

It should be a unique value for a transaction right? But in our architecture a single tx will never include multiple transfers, so I think having block number as nonce in this architecture should be fine as a single tx will never involve multiple trades, transfers or any of the operations related to TVL or Volume trading.

Comment (xsteadybcgo) [reply src/adapters/logx/execution/src/index.js:42]: I got two data entries with the same tx hash in CSV file
```
0xFfFb6B4E75c742E5DD61674a63b6D31AA3F826E6,0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870,0x2F8A25ac62179B31D62D7F80884AE57464699059,6,1,260.01300000000003,0x28e3d2ca9dcbc095a5fc63d620543d3860cd37d040640395618cdf46bb2eced2,1772423,1716057866,

0x456Ff038e9BB0F8Cb457173d4c69a57ed9642906,0x75940cDa18F14D1F97562fc2A6dBCe31CBe03870,0x2F8A25ac62179B31D62D7F80884AE57464699059,6,1,260.01300000000003,0x28e3d2ca9dcbc095a5fc63d620543d3860cd37d040640395618cdf46bb2eced2,1772423,1716057866,
```

**Pull request merged**